### ISSUE TITLE
Uable to capture RTSP on Android: Output file #0 does not contain any stream

### ISSUE THREAD

**Issue by hoonkai** [arthenica/ffmpeg-kit-test#26]
**Description**
I'm having difficulty when saving a RTSP stream into internal storage using `ffmpeg-kit -i rtsp://192.168.100.60:554/stream1 -c:v copy -c:a copy -t 5 /data/user/0/com.arthenica.ffmpegkit.test/files/test.mp4`. There're issues with the `circular_buffer_size` option, but I read that the warnings could be ignored. The command runs fine on desktop using ffmpeg. Please find below the log.

**Logs**

```
ffmpeg started on 2022-05-16 at 15:27:54
Report written to "/data/user/0/com.arthenica.ffmpegkit.test/cache/ffreport.txt"
Log level: 48
Command line:
ffmpeg-kit -i rtsp://192.168.100.60:554/stream1 -c:v copy -c:a copy -t 5 /data/user/0/com.arthenica.ffmpegkit.test/files/test.mp4
ffmpeg version v4.5-dev-3393-g30322ebe3c Copyright (c) 2000-2021 the FFmpeg developers
  built with Android (7155654, based on r399163b1) clang version 11.0.5 (https://android.googlesource.com/toolchain/llvm-project 87f1315dfbea7c137aa2e6d362dbb457e388158d)
  configuration: --cross-prefix=i686-linux-android- --sysroot=/files/android-sdk/ndk/22.1.7171670/toolchains/llvm/prebuilt/linux-x86_64/sysroot --prefix=/storage/light/projects/ffmpeg-kit/prebuilt/android-x86-lts/ffmpeg --pkg-config=/usr/bin/pkg-config --enable-version3 --arch=i686 --cpu=i686 --target-os=android --disable-neon --disable-asm --disable-inline-asm --ar=i686-linux-android-ar --cc=i686-linux-android16-clang --cxx=i686-linux-android16-clang++ --ranlib=i686-linux-android-ranlib --strip=i686-linux-android-strip --nm=i686-linux-android-nm --extra-libs='-L/storage/light/projects/ffmpeg-kit/prebuilt/android-x86-lts/cpu-features/lib -lndk_compat' --disable-autodetect --enable-cross-compile --enable-pic --enable-jni --enable-optimizations --enable-swscale --disable-static --enable-shared --enable-pthreads --enable-v4l2-m2m --disable-outdev=fbdev --disable-indev=fbdev --enable-small --disable-xmm-clobber-test --disable-debug --enable-lto --disable-neon-clobber-test --disable-programs --disable-postproc --  libavutil      57. 13.100 / 57. 13.100
  libavcodec     59. 15.102 / 59. 15.102
  libavformat    59. 10.100 / 59. 10.100
  libavdevice    59.  1.100 / 59.  1.100
  libavfilter     8. 21.100 /  8. 21.100
  libswscale      6.  1.102 /  6.  1.102
  libswresample   4.  0.100 /  4.  0.100
Splitting the commandline.
Reading option '-i' ... matched as input url with argument 'rtsp://192.168.100.60:554/stream1'.
Reading option '-c:v' ... matched as option 'c' (codec name) with argument 'copy'.
Reading option '-c:a' ... matched as option 'c' (codec name) with argument 'copy'.
Reading option '-t' ... matched as option 't' (record or transcode "duration" seconds of audio/video) with argument '5'.
Reading option '/data/user/0/com.arthenica.ffmpegkit.test/files/test.mp4' ... matched as output url.
Finished splitting the commandline.
Parsing a group of options: global .
Successfully parsed a group of options.
Parsing a group of options: input url rtsp://192.168.100.60:554/stream1.
Successfully parsed a group of options.
Opening an input file: rtsp://192.168.100.60:554/stream1.
[tcp @ 0xa21d3a80] No default whitelist set
[tcp @ 0xa21d3a80] Original list of addresses:
[tcp @ 0xa21d3a80] Address 192.168.100.60 port 554
[tcp @ 0xa21d3a80] Interleaved list of addresses:
[tcp @ 0xa21d3a80] Address 192.168.100.60 port 554
[tcp @ 0xa21d3a80] Starting connection attempt to 192.168.100.60 port 554
[tcp @ 0xa21d3a80] Successfully connected to 192.168.100.60 port 554
[rtsp @ 0xb25b5a00] SDP:
v=0
o=- 14665860 31787219 1 IN IP4 192.168.100.60
s=Session streamed by "MERCURY RTSP Server"
t=0 0
a=smart_encoder:virtualIFrame=1
m=video 0 RTP/AVP 96
c=IN IP4 0.0.0.0
b=AS:4096
a=range:npt=0-
a=control:track1
a=rtpmap:96 H264/90000
a=fmtp:96 packetization-mode=1; profile-level-id=4D4020; sprop-parameter-sets=Z01AIJWgFAF34maIAAADAAgAAAMBlCA=,aO48gA==

Failed to parse interval end specification ''
[rtsp @ 0xb25b5a00] video codec set to: h264
[rtsp @ 0xb25b5a00] RTP Packetization Mode: 1
[rtsp @ 0xb25b5a00] RTP Profile IDC: 4d Profile IOP: 40 Level: 20
[rtsp @ 0xb25b5a00] Extradata set to 0xb25abdf0 (size: 35)
[rtp @ 0xb25abe60] No default whitelist set
[udp @ 0xb25abed0] No default whitelist set
[udp @ 0xb25abed0] 'circular_buffer_size' option was set but it is not supported on this build (pthread support is required)
[udp @ 0xb25abed0] end receive buffer size reported is 524288
[udp @ 0xb25abf40] No default whitelist set
[udp @ 0xb25abf40] 'circular_buffer_size' option was set but it is not supported on this build (pthread support is required)
[udp @ 0xb25abf40] end receive buffer size reported is 524288
[rtsp @ 0xb25b5a00] setting jitter buffer size to 500
[rtsp @ 0xb25b5a00] hello state=0
Failed to parse interval end specification ''
[h264 @ 0x907e5b00] nal_unit_type: 7(SPS), nal_ref_idc: 3
[h264 @ 0x907e5b00] nal_unit_type: 8(PPS), nal_ref_idc: 3
[h264 @ 0x907e5b00] nal_unit_type: 7(SPS), nal_ref_idc: 3
[h264 @ 0x907e5b00] nal_unit_type: 8(PPS), nal_ref_idc: 3
[rtsp @ 0xb25b5a00] Could not find codec parameters for stream 0 (Video: h264, none): unspecified size
Consider increasing the value for the 'analyzeduration' (0) and 'probesize' (5000000) options
Input #0, rtsp, from 'rtsp://192.168.100.60:554/stream1':
  Metadata:
    title           : Session streamed by "MERCURY RTSP Server"
  Duration: N/A, start: 0.000000, bitrate: N/A
  Stream #0:0, 0, 1/90000: Video: h264, none, 90k tbr, 90k tbn
Successfully opened the file.
Parsing a group of options: output url /data/user/0/com.arthenica.ffmpegkit.test/files/test.mp4.
Applying option c:v (codec name) with argument copy.
Applying option c:a (codec name) with argument copy.
Applying option t (record or transcode "duration" seconds of audio/video) with argument 5.
Successfully parsed a group of options.
Opening an output file: /data/user/0/com.arthenica.ffmpegkit.test/files/test.mp4.
Output #0, mp4, to '/data/user/0/com.arthenica.ffmpegkit.test/files/test.mp4':
Output file #0 does not contain any stream
```

**Environment**

- Platform: `Android`
- Application Type: `maven central`
- Architecture: `arm64-simulator`
- Version: `v4.5.1`
- Source branch: `main`

**Comment by tanersener** [comment]
> Could not find codec parameters for stream 0 (Video: h264, none): unspecified size

Your post doesn't tell anything about which `ffmpeg-kit` package you're using. And, I see that the logs include errors about `h264` parameters. 

Make sure that you're using a package that includes an `h264` encoder or try encoding the output using another video codec.

**Comment by hoonkai** [comment]
@tanersener I just cloned the repo and ran the `android` directory. Can I ask how I can change the `ffmpeg-kit` package?

**Comment by hoonkai** [comment]
> > Could not find codec parameters for stream 0 (Video: h264, none): unspecified size
> 
> Your post doesn't tell anything about which `ffmpeg-kit` package you're using. And, I see that the logs include errors about `h264` parameters.
> 
> Make sure that you're using a package that includes an `h264` encoder or try encoding the output using another video codec.

I'm using `ffmpeg-kit-fullL4.5.1-1.LTS`.

**Comment by hoonkai** [comment]
> > Could not find codec parameters for stream 0 (Video: h264, none): unspecified size
> 
> Your post doesn't tell anything about which `ffmpeg-kit` package you're using. And, I see that the logs include errors about `h264` parameters.
> 
> Make sure that you're using a package that includes an `h264` encoder or try encoding the output using another video codec.

I'm using `ffmpeg-kit-full:4.5.1-1.LTS`. I've just tried `full-gpl`, and I get the same error. The command works fine on desktop though.

**Comment by hoonkai** [comment]
@tanersener I'm still trying to work out the cause of the issue with rtsp-simple-server, and I'm still getting the same problem. Just checking, do you get the same log as I do though after issuing the command `ffmpeg-kit -I rtsp://<ip>/<stream> -c:v copy -c:a copy -t 5 /data/user/0/com.arthenica.ffmpegkit.test/files/test.mp4`

```
ffmpeg started on 2022-06-06 at 21:32:31
    Report written to "/data/user/0/com.arthenica.ffmpegkit.test/cache/ffreport.txt"
    Log level: 48
ffmpeg version v4.5-dev-3393-g30322ebe3c
 Copyright (c) 2000-2021 the FFmpeg developers

  built with Android (7155654, based on r399163b1) clang version 11.0.5 (https://android.googlesource.com/toolchain/llvm-project 87f1315dfbea7c137aa2e6d362dbb457e388158d)
  configuration: --cross-prefix=i686-linux-android- --sysroot=/files/android-sdk/ndk/22.1.7171670/toolchains/llvm/prebuilt/linux-x86_64/sysroot --prefix=/storage/light/projects/ffmpeg-kit/prebuilt/android-x86-lts/ffmpeg --pkg-config=/usr/bin/pkg-config --enable-version3 --arch=i686 --cpu=i686 --target-os=android --disable-neon --disable-asm --disable-inline-asm --ar=i686-linux-android-ar --cc=i686-linux-android16-clang --cxx=i686-linux-android16-clang++ --ranlib=i686-linux-android-ranlib --strip=i686-linux-android-strip --nm=i686-linux-android-nm --extra-libs='-L/storage/light/projects/ffmpeg-kit/prebuilt/android-x86-lts/cpu-features/lib -lndk_compat' --disable-autodetect --enable-cross-compile --enable-pic --enable-jni --enable-optimizations --enable-swscale --disable-static --enable-shared --enable-pthreads --enable-v4l2-m2m --disable-outdev=fbdev --disable-indev=fbdev --enable-small --disable-xmm-clobber-test --disable-debug --enable-lto --disable-neon-clobber-test --disable-programs --disable-postproc --disable-doc --disable-htmlpages --disable-manpages --disable-podpages --disable-txtpages --disable-sndio --disable-schannel --disable-securetransport --disable-xlib --disable-cuda --disable-cuvid --disable-nvenc --disable-vaapi --disable-vdpau --disable-videotoolbox --disable-audiotoolbox --disable-appkit --disable-alsa --disable-cuda --disable-cuvid --disable-nvenc --disable-vaapi --disable-vdpau --enable-libfontconfig --enable-libfreetype --enable-libfribidi --enable-gmp --enable-gnutls --enable-libmp3lame --enable-libass --enable-iconv --enable-libtheora --enable-libvorbis --enable-libvpx --enable-libwebp --enable-libxml2 --enable-libopencore-amrnb --enable-libshine --enable-libspeex --enable-libdav1d --enable-libkvazaar --enable-libx264 --enable-libxvid --enable-libx265 --enable-libvidstab --enable-libilbc --enable-libopus --enable-libsnappy --enable-libsoxr --enable-libtwolame --disable-sdl2 --enable-libvo-amrwbenc --enable-libzimg --disable-openssl --enable-zlib --enable-mediacodec --enable-gpl
  libavutil      57. 13.100 / 57. 13.100
  libavcodec     59. 15.102 / 59. 15.102
  libavformat    59. 10.100 / 59. 10.100
  libavdevice    59.  1.100 / 59.  1.100
  libavfilter     8. 21.100 /  8. 21.100
  libswscale      6.  1.102 /  6.  1.102
  libswresample   4.  0.100 /  4.  0.100
[udp @ 0xf45f4910] 'circular_buffer_size' option was set but it is not supported on this build (pthread support is required)
[udp @ 0xf45f5990] 'circular_buffer_size' option was set but it is not supported on this build (pthread support is required)
[udp @ 0xf45df490] 'circular_buffer_size' option was set but it is not supported on this build (pthread support is required)
[udp @ 0xf45df310] 'circular_buffer_size' option was set but it is not supported on this build (pthread support is required)
```

**Comment by hoonkai** [comment]
I've tried it on a physical phone and the file is written. However, the command wouldn't terminate even when `-t N` was used. Is ffmpeg-kit supposed to work in an emulator?

**Comment by tanersener** [comment]
I spent quite some time analysing this issue yesterday. Reviewed my previous tests with `rtsp-simple-server` and fixed a few problems I had before. 

Long story short, `UDP` streaming doesn't work on `Android` and there is no fix for that. It is not something we can implement easily. `Android`'s `pthread` implementation is to blame there. I explained that on [Tips 8.](https://github.com/tanersener/ffmpeg-kit/wiki/Tips#8-udp-streaming-support-on-android). 

The only available alternative is to use `tcp` instead of `udp`. In your case, if you add `-rtsp_transport tcp` to your command it will work as expected.

**Comment by hoonkai** [comment]
> I spent quite some time analysing this issue yesterday. Reviewed my previous tests with `rtsp-simple-server` and fixed a few problems I had before.
> 
> Long story short, `UDP` streaming doesn't work on `Android` and there is no fix for that. It is not something we can implement easily. `Android`'s `pthread` implementation is to blame there. I explained that on [Tips 8.](https://github.com/tanersener/ffmpeg-kit/wiki/Tips#8-udp-streaming-support-on-android).
> 
> The only available alternative is to use `tcp` instead of `udp`. In your case, if you add `-rtsp_transport tcp` to your command it will work as expected.

It does produce a video + audio capture on a physical phone though - it just wouldn't quit even when `-t 200` is used. Does `ffmpeg-kit` respect the `-t` option?

**Comment by tanersener** [comment]
Of course it does. I didn't have any problems about it when I used the `rtsp_transport tcp` option in my tests. 

I also used a physical device but the behaviour didn't change for me. 

As I explained above, there are some `pthreads` functions not available on `Android` that are used by the `udp` module. I believe they are the reason why the commands are not terminated properly. Not that `ffmpeg-kit` doesn't respect `-t` option.
